cubedash/testutils/database.py

```diff
@@ -19,129 +20,252 @@
-        # to enable this fixture to not be function scoped
+        # so that this fixture can be module scoped and not function scoped.
         mp = pytest.MonkeyPatch()
 
+        # Share the temporary configuration file with the tests
         mp.setenv("ODC_CONFIG_PATH", str(temp_datacube_config_file.absolute()))
-        yield postgres_url
+        yield new_postgresql_url
         mp.undo()
 
+        # This was a good idea to save disk space on the tmpfs
+        # But something is holding connections open to the database
+        # E  psycopg2.errors.ObjectInUse: database "test_dataset_maturity_default" is being accessed
+        # by other users
+        # E  DETAIL:  There are 2 other sessions using the database.
+        # conn = psycopg2.connect(postgresql_server)
+        # # Prevent starting a transaction, since 'drop database' cannot run inside a transaction
+        # conn.set_session(autocommit=True)
+        # with conn.cursor() as cur:
+        #     log.info("Dropping database", db_name=test_database_name)
+        #     cur.execute(
+        #         sql.SQL("DROP DATABASE {db_name}").format(
+        #             db_name=sql.Identifier(test_database_name)
+        #         )
+        #     )
+        # conn.close()
+
 
 @pytest.fixture(scope="module", params=["default", "postgis"])
-def env_name(request) -> str:
+def env_name(request: pytest.FixtureRequest) -> str:
     return request.param
 
 
 @pytest.fixture(scope="module")
-def cfg_env(odc_db, env_name) -> ODCEnvironment:
-    """Provides a :class:`ODCEnvironment` configured with suitable config file paths."""
-    return ODCConfig()[env_name]
+def cfg_env(odc_db, env_name: str) -> ODCEnvironment:
+    """Provide a :class:`ODCEnvironment` configured with suitable config file paths."""
+    # Need to request `odc_db` since it exports an environment variable specifying which
+    # configuration file to use. So this fixture *must* run after that fixture.
+    odc_env = ODCConfig().get_environment(env_name)
+    assert odc_env.db_url is not None
+    logger.info(
+        "Prepping ODCConfig",
+        env_name=env_name,
+        odc_env_url=odc_env.db_url,
+    )
+    return odc_env
 
 
 @pytest.fixture(scope="module")
-def odc_test_db(cfg_env):
-    """
-    Provide a temporary PostgreSQL server initialised by ODC, usable as
-    the default ODC DB by setting environment variables.
-    :return: Datacube instance
-    """
+def odc_test_db(odc_db: str, cfg_env: ODCEnvironment) -> Generator[Datacube]:
+    """Provide a temporary PostgreSQL server initialised by ODC.
+
+    Usable as the default ODC DB by setting environment variables.
+
+    Yields:
+        Ready to go Datacube Object
 
+    """
+    logger.info(
+        "Initialising ODC Schema for testing", odc_db=odc_db, cfg_db_url=cfg_env.db_url
+    )
     index = index_connect(cfg_env, validate_connection=False)
+    # With permissions causes trouble when run multiple times on the same database
```

Comment:
I'm not sure what problem you're seeing, but could you be hitting https://github.com/opendatacube/datacube-core/issues/1959?